ddtrace/llmobs/_llmobs.py

```diff
@@ -774,6 +777,141 @@ def _tag_metrics(span: Span, metrics: Dict[str, Any]) -> None:
+
+        has_exactly_one_joining_key = (span is not None) ^ (span_with_tag is not None)
+
+        if not has_exactly_one_joining_key:
+            raise ValueError(
+                "Exactly one of `span` or `span_with_tag` must be specified to submit an evaluation metric."
+            )
+
+        join_on = {}
+        if span is not None:
+            if (
+                not isinstance(span, dict)
+                or "span_id" not in span
+                or "trace_id" not in span
+                or not isinstance(span.get("span_id"), str)
+                or not isinstance(span.get("trace_id"), str)
+            ):
+                raise TypeError(
+                    "`span` must be a dictionary containing both span_id and trace_id keys. "
+                    "LLMObs.export_span() can be used to generate this dictionary from a given span."
+                )
+            join_on["span"] = span
+        elif span_with_tag is not None:
+            if (
+                not isinstance(span_with_tag, tuple)
+                or len(span_with_tag) != 2
+                or not all(isinstance(i, str) for i in span_with_tag)
+            ):
+                raise TypeError("`span_with_tag` must be a tuple of shape (tag_key, tag_value)")
+            join_on["tag"] = {"tag_key": span_with_tag[0], "tag_value": span_with_tag[1]}
+
+        timestamp_ms = timestamp_ms if timestamp_ms else int(time.time() * 1000)
+
+        if not isinstance(timestamp_ms, int) or timestamp_ms < 0:
+            raise ValueError("timestamp_ms must be a non-negative integer. Evaluation metric data will not be sent")
+
+        if not label:
+            raise ValueError("label must be the specified name of the evaluation metric.")
+
+        metric_type = metric_type.lower()
+        if metric_type not in ("categorical", "score"):
+            raise ValueError("metric_type must be one of 'categorical' or 'score'.")
+
+        if metric_type == "categorical" and not isinstance(value, str):
+            raise TypeError("value must be a string for a categorical metric.")
+        if metric_type == "score" and not isinstance(value, (int, float)):
+            raise TypeError("value must be an integer or float for a score metric.")
+
+        if tags is not None and not isinstance(tags, dict):
+            log.warning("tags must be a dictionary of string key-value pairs.")
+            tags = None
+
+        # initialize tags with default values that will be overridden by user-provided tags
+        evaluation_tags = {
+            "ddtrace.version": ddtrace.__version__,
+            "ml_app": ml_app,
+        }
+
+        if tags:
+            for k, v in tags.items():
+                try:
+                    evaluation_tags[ensure_text(k)] = ensure_text(v)
+                except TypeError:
+                    log.warning("Failed to parse tags. Tags for evaluation metrics must be strings.")
+
+        if not config._dd_api_key:
+            log.warning(
```

Comment:
we should get this warning later on when the writer tries to send without an api key right? Just want to make sure we're not going to double log for this

Reply:
We have a generic error log for when we fail to send in the eval metric writer, when there's no api key we'll show an auth error
```
        logger.error(
            "failed to send %d LLMObs %s events to %s", len(events), self._event_type, self._intake, exc_info=True
        )
```
Instead of relying on this error log, we can do a check in `periodic` for the eval metric writer for the existence of an api key and log a more friendly warning message, then return

```
        if not self._headers.get("DD-API-KEY"):
            logger.warning(
                "DD_API_KEY is required for sending evaluation metrics. Evaluation metric data will not be sent. ",
                "Ensure this configuration is set before running your application."
            )
            return
```